[BEAM-2740] introduce the new InnerBeamSqlEnv/BeamSqlEnv

xumingming opened the pull request:
Summary:

* Simplified `BeamSqlEnv`: now you can only register UDF/UDAF in `BeamSqlEnv`, and `BeamSqlEnv` just record the udfName -> udfClass mapping, actually work will be done by `InnerBeamSqlEnv`.
    * `BeamSqlEnv` is only used by the surface API classes.
* `InnerBeamSqlEnv` is actually the old `BeamSqlEnv`, it holds schema and planner.
  * `InnerBeamSqlEnv` is only used by the implementation classes.
* `BeamSqlEnv` is converted to `InnerBeamSqlEnv` automatically in `BeamSql` and `BeamSqlCli`.
* `BeamSqlEnv` and `InnerBeamSqlEnv` are quite different, so can not share the same interface.

xumingming commented:
R: @takidau @XuMingmin

takidau commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/BeamSqlEnv.java:
```diff
@@ -18,111 +18,64 @@
- * <p>It contains a {@link SchemaPlus} which holds the metadata of tables/UDF functions, and
- * a {@link BeamQueryPlanner} which parse/validate/optimize/translate input SQL queries.
+ * <p>This class mainly provides methods to register {@code UDF} and {@code UDAF}, which can be
+ * used
```
Did you mean to write more at the end of this comment?

xumingming replied:
Yeah... Will add the full javadoc.

takidau commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/BeamSqlEnv.java:
```diff
@@ -18,111 +18,64 @@
-    planner.getSourceTables().put(tableName, table);
+  private void validateUdfNameUniqueness(String name) {
+    if (udfs.containsKey(name) || udafs.containsKey(name) || sfnUdfs.containsKey(name)) {
+      throw new IllegalArgumentException("UDF named: " + name + " is already registered!");
```
Do we really want this check? I thought we were planning on allowing overrides to registrations. This will be particularly useful once we have service loader registration. You might automatically register a version of a UDF/UDAF that you don't actually want to use, in which case we want to provide a way to override that.

xumingming replied:
I was thinking that auto-override might cause issues, e.g. user have defined a UDF named `myudf`, but an third-party library might register the same name which override it quietly, while user don't know that?

XuMingmin replied:
The check-fail pattern looks good to me, to avoid unexpected function conflict.

takidau replied:
Okay, let's go with it for now. My concern is that users will find they have no way to override UDF registrations, but also will not have the option of not including some auto-registered library UDF that they don't want to use. Probably okay to cross that bridge when we get to it.

takidau commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/InnerBeamSqlEnv.java:
```diff
@@ -0,0 +1,164 @@
+ * holds the metadata of tables/UDF functions, and a {@link BeamQueryPlanner} which
+ * parse/validate/optimize/translate input SQL queries.
+ */
+public class InnerBeamSqlEnv implements Serializable{
```
The "Inner" prefix makes me think this might be an inner class, which it isn't. Given that it lives in an impl package, would BeamSqlEnvImpl be a better name? BeamSqlEnvInternal?

xumingming replied:
`BeamSqlEnvImpl` seems not good, since `BeamSqlEnvImpl` is not a implementation of `BeamSqlEnv`; `BeamSqlEnvInternal` sounds ok to me.

xumingming replied:
or `InternalBeamSqlEnv`?

takidau replied:
My vote is @XuMingmin's suggestion below.

takidau commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/BeamSqlEnv.java:
```diff
@@ -18,111 +18,64 @@
+public class BeamSqlEnv implements Serializable {
+  private Map<String, Class<? extends BeamSqlUdf>> udfs = new HashMap<>();
+  private Map<String, Class<? extends BeamSqlUdaf>> udafs = new HashMap<>();
+  private Map<String, SerializableFunction> sfnUdfs = new HashMap<>();
```
I feel like it would be conceptually cleaner to just have an InnerBeamSqlEnv object owned by the BeamSqlEnv that all register calls would get immediately passed down to. There could then be a getImpl() method clearly marked as being a non-public API that could return the InnerBeamSqlEnv object used in the implementation. I worry that, as currently written, it's possible for the BeamSqlEnv and any InnerBeamSqlEnv created from it to diverge, and that may be non-obvious. What do you think?

xumingming replied:
I understand your concern. By "marked as being a non-public API", you mean use some kind of annotation? Then the `InnerBeamSqlEnv` will be exposed to end-user, will the annotation in question handle this exposure?

takidau replied:
I was just thinking a comment, but I think @XuMingmin's comment below about just moving BeamSqlEnv into the impl package makes the whole thing moot. That seems like a better route overall.

takidau commented:
I like that suggestion.

xumingming commented:
BeamSqlEnv can be removed if we only consider BeamSql, but it is needed very soon for the BeamSqlCli(I already have a pending PR: #3481 ), I'd prefer to make it ready for the comming BeamSqlCli PR.

XuMingmin commented:
I guess `BeamSqlEnv` can be hidden(I suppose any class in package `*.impl` is treated as hidden) for both `BeamSql` and `BeamSqlCli`. 
Each `BeamSqlCli` will have a shared `BeamSqlEnv`, and `public void execute(String sqlString)` would be the only exposed method, --TBD in future regarding to details of `BeamSqlCli`.

xumingming commented:
Rebased.

Summary:

* Only keep one `BeamSqlEnv` in `impl` package.
* Re-organized the surfacing apis to only remain: 
  * BeamSql
  * BeamSqlUdf
  * BeamRecordSqlType
  * BeamSqlRecordHelper 
* Refactored `QueryTransform` and `SimpleQueryTransform` to enable more code reuse.

takidau commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/BeamSql.java:
```diff
@@ -121,72 +113,67 @@ public static SimpleQueryTransform simpleQuery(String sqlQuery) {
     //register tables, related with input PCollections.
-    private void registerTables(PCollectionTuple input){
+    private void registerTables(PCollectionTuple input,
+        BeamSqlEnv beamSqlEnv){
```
Why is the parameter needed? Can't it just use the member variable?

xumingming replied:
Not needed.

takidau commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/BeamSql.java:
```diff
@@ -197,53 +184,47 @@ private void registerTables(PCollectionTuple input){
     private static final String PCOLLECTION_TABLE_NAME = "PCOLLECTION";
-    abstract BeamSqlEnv getSqlEnv();
-    abstract String getSqlQuery();
+    private QueryTransform delegate;
```
Nice.

takidau commented on sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamJoinRelBoundedVsBoundedTest.java:
```diff
@@ -34,10 +33,10 @@
   @Rule
   public final TestPipeline pipeline = TestPipeline.create();
-  private static final BeamSqlEnv beamSqlEnv = new BeamSqlEnv();
+  private static final BeamSqlEnv INNER_BEAM_SQL_ENV = new BeamSqlEnv();
```
Update the name, now that it's no longer named "inner"?

takidau commented on sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamJoinRelUnboundedVsBoundedTest.java:
```diff
@@ -39,18 +38,18 @@
   @Rule
   public final TestPipeline pipeline = TestPipeline.create();
-  private static final BeamSqlEnv beamSqlEnv = new BeamSqlEnv();
+  private static final BeamSqlEnv INNER_BEAM_SQL_ENV = new BeamSqlEnv();
```
Update name?

takidau commented on sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamJoinRelUnboundedVsUnboundedTest.java:
```diff
@@ -38,18 +37,18 @@
   @Rule
   public final TestPipeline pipeline = TestPipeline.create();
-  private static final BeamSqlEnv beamSqlEnv = new BeamSqlEnv();
+  private static final BeamSqlEnv INNER_BEAM_SQL_ENV = new BeamSqlEnv();
```
Update name?

takidau commented:
Merged, feel free to close.
